api/trace/testtrace/span.go

```diff
@@ -0,0 +1,213 @@
+
+	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/api/trace"
+)
+
+var _ trace.Span = (*Span)(nil)
+
+type Span struct {
+	lock         *sync.Mutex
+	tracer       *Tracer
+	spanContext  core.SpanContext
+	parentSpanID core.SpanID
+	ended        bool
+	name         string
+	startTime    time.Time
+	endTime      time.Time
+	status       codes.Code
+	attributes   map[core.Key]core.Value
+	events       []Event
+	links        map[core.SpanContext][]core.KeyValue
+}
+
+func (s *Span) Tracer() trace.Tracer {
+	return s.tracer
+}
+
+func (s *Span) End(opts ...trace.EndOption) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.ended {
+		return
+	}
+
+	var c trace.EndOptions
+
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	s.endTime = time.Now()
+
+	if endTime := c.EndTime; !endTime.IsZero() {
+		s.endTime = endTime
+	}
+
+	s.ended = true
+}
+
+func (s *Span) AddEvent(ctx context.Context, msg string, attrs ...core.KeyValue) {
+	s.AddEventWithTimestamp(ctx, time.Now(), msg, attrs...)
+}
+
+func (s *Span) AddEventWithTimestamp(ctx context.Context, timestamp time.Time, msg string, attrs ...core.KeyValue) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.ended {
+		return
+	}
+
+	attributes := make(map[core.Key]core.Value)
+
+	for _, attr := range attrs {
+		attributes[attr.Key] = attr.Value
+	}
+
+	s.events = append(s.events, Event{
+		Timestamp:  timestamp,
+		Message:    msg,
+		Attributes: attributes,
+	})
+}
+
+func (s *Span) IsRecording() bool {
+	return true
+}
+
+func (s *Span) SpanContext() core.SpanContext {
+	return s.spanContext
+}
+
+func (s *Span) SetStatus(status codes.Code) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.ended {
+		return
+	}
+
+	s.status = status
+}
+
+func (s *Span) SetName(name string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.ended {
+		return
+	}
+
+	s.name = name
+}
+
+func (s *Span) SetAttribute(attr core.KeyValue) {
+	s.SetAttributes(attr)
+}
+
+func (s *Span) SetAttributes(attrs ...core.KeyValue) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.ended {
+		return
+	}
+
+	for _, attr := range attrs {
+		s.attributes[attr.Key] = attr.Value
+	}
+}
+
+// Name returns the name most recently set on the Span, either at or after creation time.
+// It cannot be change after End has been called on the Span.
+func (s *Span) Name() string {
+	return s.name
+}
+
+// ParentSpanID returns the SpanID of the parent Span.
+// If the Span is a root Span and therefore does not have a parent, the returned SpanID will be invalid
+// (i.e., it will contain all zeroes).
+func (s *Span) ParentSpanID() core.SpanID {
+	return s.parentSpanID
+}
+
+// Attributes returns the attributes set on the Span, either at or after creation time.
+// If the same attribute key was set multiple times, the last call will be used.
+// Attributes cannot be changed after End has been called on the Span.
+func (s *Span) Attributes() map[core.Key]core.Value {
+	attributes := make(map[core.Key]core.Value)
```

Comment:
Should these accessors take the lock? The race detector could notice that `SetAttributes()` takes a lock and `Attributes()` does not under some circumstances, I imagine.

Reply:
Added the lock + a test that triggers the race detector

(The race detector isn't currently able to notice symmetrical operations like getting/setting, but actually _running_ both the getter and setter concurrently will trigger it.)